File: update-imports.py
#!/usr/bin/env python3
"""
Update all import statements after file renaming.
This script is more reliable than sed-based approaches.
"""

import os
import re
import json
from pathlib import Path
from typing import Dict, List, Tuple

# Load the rename mapping
SCRIPT_DIR = Path(__file__).parent
MAPPING_FILE = SCRIPT_DIR / 'rename-mapping.json'

def load_mappings() -> List[Tuple[str, str]]:
    """Load rename mappings from JSON file."""
    with open(MAPPING_FILE, 'r') as f:
        data = json.load(f)

    mappings = []
    for category in data['rename_mappings'].values():
        if isinstance(category, list):
            for item in category:
                old_path = item['from'].replace('src/', '')
                new_path = item['to'].replace('src/', '')
                # Remove file extensions for matching
                old_path = re.sub(r'\.(tsx?|jsx?)$', '', old_path)
                new_path = re.sub(r'\.(tsx?|jsx?)$', '', new_path)
                mappings.append((old_path, new_path))

    return mappings

def update_import_in_line(line: str, mappings: List[Tuple[str, str]]) -> str:
    """Update import statement in a single line."""
    original_line = line

    for old_path, new_path in mappings:
        old_file = os.path.basename(old_path)
        new_file = os.path.basename(new_path)
        old_dir = os.path.dirname(old_path)
        new_dir = os.path.dirname(new_path)

        # Pattern 1: Exact full path match
        # from './components/OldFile' or from '../components/OldFile'
        pattern1 = rf"(from\s+['\"`])(\./|\.\./)({re.escape(old_path)})(['\"`])"
        replacement1 = rf"\1\2{new_path}\4"
        line = re.sub(pattern1, replacement1, line)

        # Pattern 2: File name only (when in same/parent dir)
        # from './OldFile'
        if old_file != new_file:
            pattern2 = rf"(from\s+['\"`](?:\./|\.\./)?)({re.escape(old_file)})(['\"`])"
            replacement2 = rf"\1{new_file}\3"
            line = re.sub(pattern2, replacement2, line)

        # Pattern 3: Alias imports (if using @/ or similar)
        pattern3 = rf"(from\s+['\"`]@/)({re.escape(old_path)})(['\"`])"
        replacement3 = rf"\1{new_path}\3"
        line = re.sub(pattern3, replacement3, line)

    return line

def process_file(file_path: Path, mappings: List[Tuple[str, str]]) -> int:
    """Process a single file and update imports."""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            lines = f.readlines()

        updated_lines = []
        changes = 0

        for line in lines:
            updated_line = update_import_in_line(line, mappings)
            if updated_line != line:
                changes += 1
            updated_lines.append(updated_line)

        if changes > 0:
            with open(file_path, 'w', encoding='utf-8') as f:
                f.writelines(updated_lines)

        return changes
    except Exception as e:
        print(f"Error processing {file_path}: {e}")
        return 0

def find_typescript_files(src_dir: Path) -> List[Path]:
    """Find all TypeScript and JavaScript files."""
    extensions = ['.ts', '.tsx', '.js', '.jsx']
    files = []

    for ext in extensions:
        files.extend(src_dir.rglob(f'*{ext}'))

    return [f for f in files if 'node_modules' not in str(f)]

def main():
    """Main execution."""
    print("=== Import Update Script (Python) ===\n")

    # Load mappings
    print("Loading rename mappings...")
    mappings = load_mappings()
    print(f"Found {len(mappings)} file renames\n")

    # Find all TypeScript files
    src_dir = SCRIPT_DIR / 'src'
    if not src_dir.exists():
        print(f"Error: src directory not found at {src_dir}")
        return 1

    print("Finding TypeScript files...")
    files = find_typescript_files(src_dir)
    print(f"Found {len(files)} files to process\n")

    # Process each file
    print("Updating imports...")
    total_changes = 0
    files_changed = []

    for file_path in files:
        changes = process_file(file_path, mappings)
        if changes > 0:
            total_changes += changes
            files_changed.append(file_path)
            relative_path = file_path.relative_to(SCRIPT_DIR)
            print(f"  ✓ {relative_path} ({changes} change{'s' if changes != 1 else ''})")

    # Summary
    print(f"\n{'='*50}")
    print("Summary:")
    print(f"  Files processed: {len(files)}")
    print(f"  Files modified: {len(files_changed)}")
    print(f"  Total changes: {total_changes}")

    if files_changed:
        print(f"\n{'='*50}")
        print("Modified files:")
        for file_path in files_changed:
            print(f"  - {file_path.relative_to(SCRIPT_DIR)}")

    print(f"\n{'='*50}")
    if total_changes > 0:
        print("✅ Import updates complete!")
        print("\nNext steps:")
        print("  1. Run: npm run type-check")
        print("  2. Run: npm run build")
        print("  3. Review changes: git diff")
    else:
        print("ℹ️  No import statements needed updating.")

    return 0

if __name__ == '__main__':
    exit(main())

File: update-imports.js
#!/usr/bin/env node

/**
 * Script to update all import statements after file renaming
 * This script reads the rename-mapping.json and updates all imports accordingly
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

// Read the rename mapping
const mappingPath = path.join(__dirname, 'rename-mapping.json');
const mapping = JSON.parse(fs.readFileSync(mappingPath, 'utf8'));

// Build a map of old paths to new paths
const renameMap = new Map();

// Process all categories in the mapping
Object.values(mapping.rename_mappings).forEach(category => {
  if (Array.isArray(category)) {
    category.forEach(rename => {
      // Remove 'src/' prefix and file extension for matching
      const fromPath = rename.from.replace(/^src\//, '').replace(/\.(tsx?|jsx?)$/, '');
      const toPath = rename.to.replace(/^src\//, '').replace(/\.(tsx?|jsx?)$/, '');
      renameMap.set(fromPath, toPath);
    });
  }
});

console.log('=== Import Update Script ===');
console.log(`Found ${renameMap.size} file renames to process`);
console.log('');

// Function to convert PascalCase path to kebab-case
function toKebabCase(str) {
  return str
    .replace(/([A-Z])/g, '-$1')
    .toLowerCase()
    .replace(/^-/, '');
}

// Build sed commands for each rename
const sedCommands = [];

renameMap.forEach((newPath, oldPath) => {
  // Handle different import patterns:
  // 1. from './OldFile' or from '../OldFile' (relative imports)
  // 2. from '@/OldFile' (alias imports)
  // 3. from 'components/OldFile' (direct imports)

  const oldFileName = path.basename(oldPath);
  const newFileName = path.basename(newPath);
  const oldDir = path.dirname(oldPath);
  const newDir = path.dirname(newPath);

  // Create sed pattern for this rename
  // This will match: from './OldFile', from '../path/OldFile', from '@/path/OldFile'
  if (oldFileName !== newFileName || oldDir !== newDir) {
    sedCommands.push({
      old: oldPath,
      new: newPath,
      oldFile: oldFileName,
      newFile: newFileName
    });
  }
});

console.log('Building find & replace patterns...');
console.log('');

// Create a Node script to do the replacements
const replacements = Array.from(renameMap.entries()).map(([oldPath, newPath]) => {
  return { oldPath, newPath };
});

const scriptContent = `
const fs = require('fs');
const path = require('path');
const glob = require('glob');

const replacements = ${JSON.stringify(replacements, null, 2)};

// Find all TypeScript and JavaScript files
const files = glob.sync('src/**/*.{ts,tsx,js,jsx}', {
  cwd: '${process.cwd()}',
  absolute: true
});

console.log(\`Processing \${files.length} files...\`);

let totalReplacements = 0;
const filesChanged = new Set();

files.forEach(file => {
  let content = fs.readFileSync(file, 'utf8');
  let fileChanged = false;

  replacements.forEach(({ oldPath, newPath }) => {
    const oldFile = oldPath.split('/').pop();
    const newFile = newPath.split('/').pop();

    // Pattern 1: Match exact path in imports (handles ./, ../, @/)
    const pattern1 = new RegExp(
      \`(from\\\\s+['"\\\`])(\\\\.\\\\.?/.*/)?\${oldPath.replace(/\\//g, '\\\\/')}(['"\\\`])\`,
      'g'
    );

    // Pattern 2: Match filename only (for when directory is already correct)
    const pattern2 = new RegExp(
      \`(from\\\\s+['"\\\`])(\\\\.\\\\.?/.*/)?\${oldFile}(['"\\\`])\`,
      'g'
    );

    const newContent1 = content.replace(pattern1, (match, before, dir, after) => {
      totalReplacements++;
      fileChanged = true;
      return before + (dir || '') + newPath + after;
    });

    const newContent2 = newContent1.replace(pattern2, (match, before, dir, after) => {
      if (oldFile !== newFile) {
        totalReplacements++;
        fileChanged = true;
        return before + (dir || '') + newFile + after;
      }
      return match;
    });

    content = newContent2;
  });

  if (fileChanged) {
    fs.writeFileSync(file, content, 'utf8');
    filesChanged.add(file);
  }
});

console.log(\`\\nComplete!\`);
console.log(\`Files changed: \${filesChanged.size}\`);
console.log(\`Total replacements: \${totalReplacements}\`);

if (filesChanged.size > 0) {
  console.log('\\nFiles modified:');
  Array.from(filesChanged).forEach(f => {
    console.log(\`  - \${f.replace(process.cwd() + '/', '')}\`);
  });
}
`;

// Write the update script
const updateScriptPath = path.join(__dirname, 'update-imports-worker.js');
fs.writeFileSync(updateScriptPath, scriptContent, 'utf8');

console.log('Update script created at: update-imports-worker.js');
console.log('');
console.log('To run the import updates:');
console.log('  npm install glob  # If not already installed');
console.log('  node update-imports-worker.js');
console.log('');
console.log('Or use your IDE\'s refactoring feature for automatic updates!');
